### ISSUE TITLE
Restore removed option for disabling intermediate simplify()

### ISSUE THREAD

**Issue by clinssen** [sympy/sympy#19532]
Since upgrading to sympy==1.6 from sympy==1.5, we noticed that our CI tests started failing due to timeout. A matrix exponential on a simple 7x7 matrix had become about 10x slower, causing the issue. Git bisect on sympy tracked it down to:

#18572 Enabling matrix intermediate product simplification by default

The option to disable intermediate calls to simplify() was removed. It is not clear to me why this option was removed (in addition to just having the default changed). @oscarbenjamin already mentioned the performance regression in matrix powers in that thread (https://github.com/sympy/sympy/pull/18572#issuecomment-621810132) but it was not addressed. Expression simplification is a bottleneck in our application because of the length of our expressions, and we need to use simplification as sparingly as possible.

Please re-instate the option to disable intermediate calls to expression simplification.

@Pristine-Cat @oscarbenjamin I can start a PR implementing these changes, just give me a heads up.

**Comment by oscarbenjamin** [comment]
Note that it is possible to disable simplification by using the pow method to control how you want the power to be evaluated:
```julia
In [21]: M = Matrix([[1+I, 1-I], [1-I, 1+I]])                                                                                     

In [22]: M                                                                                                                        
Out[22]: 
⎡1 + ⅈ  1 - ⅈ⎤
⎢            ⎥
⎣1 - ⅈ  1 + ⅈ⎦

In [23]: M ** 2                                                                                                                   
Out[23]: 
⎡0  4⎤
⎢    ⎥
⎣4  0⎦

In [24]: M.pow(2, method='multiply')                                                                                              
Out[24]: 
⎡       2          2                     ⎤
⎢(1 - ⅈ)  + (1 + ⅈ)    2⋅(1 - ⅈ)⋅(1 + ⅈ) ⎥
⎢                                        ⎥
⎢                            2          2⎥
⎣ 2⋅(1 - ⅈ)⋅(1 + ⅈ)   (1 - ⅈ)  + (1 + ⅈ) ⎦
```
The idea is that it if you particularly care about controlling how the power is calculated (as it seems you do) then it is better to be able to explicitly select the method you want rather than depending on any particular default method.

I presume that you are referring to the global flag. There is an environment variable that controls this so if you set `SYMPY_DOTPRODSIMP=0` then that will also disable it:
```julia
In [1]: M = Matrix([[1+I, 1-I], [1-I, 1+I]])                                                                                      

In [2]: M ** 2                                                                                                                    
Out[2]: 
⎡       2          2                     ⎤
⎢(1 - ⅈ)  + (1 + ⅈ)    2⋅(1 - ⅈ)⋅(1 + ⅈ) ⎥
⎢                                        ⎥
⎢                            2          2⎥
⎣ 2⋅(1 - ⅈ)⋅(1 + ⅈ)   (1 - ⅈ)  + (1 + ⅈ) ⎦
```
We don't want to have that environment variable though so don't expect that to be there in the long term. I think the idea was that it should be there as a quick fix just in case anyone experienced a regression with these changes.

**Comment by clinssen** [comment]
Setting the ``SYMPY_DOTPRODSIMP`` global does indeed yield the original behaviour I was looking for, thanks for pointing me to this! 

If it will be going away soon, could you perhaps hint at how the calls to sympy.exp() should be changed if the argument is a matrix? exp() does not seem to accept the "method" argument.

**Comment by oscarbenjamin** [comment]
Can you show an example of the matrix you are using? It might be useful for benchmarks.

The `Matrix.exp` method does not seem to have an option for this. It would be a reasonable PR to add method arguments for `Matrix.exp` similar to `pow`. The other method that could be added there is cayley based on `M.analytic_func(exp(x), x)` which can also be used for computing the matrix exponential.

**Comment by clinssen** [comment]
This is an actual matrix that comes up in our application:

```Python
import sympy

a, b, c, d, e = sympy.symbols("a b c d e")

M = sympy.Matrix([
[      0,  1.0,       0,    0,       0,    0,    0],
[-1/a**2, -2/a,       0,    0,       0,    0,    0],
[      0,    0,       0,  1.0,       0,    0,    0],
[    0.0,  0.0, -1/b**2, -2/b,       0,    0,    0],
[    0.0,  0.0,     0.0,  0.0,     0.0,    1,    0],
[    0.0,  0.0,     0.0,  0.0, -1/c**2, -2/c,    0],
[    1/d,  0.0,     1/d,  0.0,     1/d,  0.0, -1/e]])
```
Using sympy 1.5:
```Python
>>> %timeit -n1 sympy.exp(M)
1 loop, best of 3: 1min per loop
```
Using sympy 1.6: left running overnight, found killed in the morning due to out of memory on the same system (i7-8550U/16 GB RAM)

**Comment by oscarbenjamin** [comment]
That's an interesting example since most 7x7 matrices can not have an ordinary analytic form for the matrix exponential due to Abel-Ruffini. This matrix (ignoring the spurious floats) has coefficients in `QQ(a, b, c, d, e)` though and the characteristic polynomial factors completely in that field:
```julia
In [4]: M.applyfunc(nsimplify).eigenvals()                                                                                        
Out[4]: 
⎧-1      -1      -1      -1    ⎫
⎨───: 2, ───: 2, ───: 2, ───: 1⎬
⎩ a       b       c       e    ⎭
```
That should mean that almost all of the calculations here can be done in `QQ(a, b, c, d, e)` and indeed the answer does not seem to be necessarily complicated.

What is slow here is actually `rref` used a part of computing the eigenvectors. The intermediate term simplification slows this down by using `cancel`. It seems that the call to `cancel` is generally the culprit in cases where intermediate term simplification slows things down. Where cancellation occurs the use of `cancel` can speed things up enormously but the flipside is that it can also slows things down enormously when there is no possible cancellation.

Since the matrix is defective in order to compute the Jordan form we need to compute the null space of `(M - I lambda)**2` which is a lot slower than for `(M - I lambda)` so the non-trivial Jordan blocks are relevant to the slowness.

With sympy 1.5 we have
```julia
In [6]: M2 = (M.applyfunc(nsimplify) + eye(7)/a) ** 2                                                                                          

In [7]: M2                                                                                                                                     
Out[7]: 
⎡      0        0         0                 0                0                 0             0     ⎤
⎢                                                                                                  ⎥
⎢      0        0         0                 0                0                 0             0     ⎥
⎢                                                                                                  ⎥
⎢                       1    1             2   2                                                   ⎥
⎢      0        0     - ── + ──          - ─ + ─             0                 0             0     ⎥
⎢                        2    2            b   a                                                   ⎥
⎢                       b    a                                                                     ⎥
⎢                                                                                                  ⎥
⎢                      2   1                                                                       ⎥
⎢                    - ─ + ─                  2                                                    ⎥
⎢                      b   a    1    ⎛  2   1⎞    1                                                ⎥
⎢      0        0  - ─────── - ────  ⎜- ─ + ─⎟  - ──         0                 0             0     ⎥
⎢                        2        2  ⎝  b   a⎠     2                                               ⎥
⎢                       b      a⋅b                b                                                ⎥
⎢                                                                                                  ⎥
⎢                                                          1    1             2   2                ⎥
⎢      0        0         0                 0            - ── + ──          - ─ + ─          0     ⎥
⎢                                                           2    2            c   a                ⎥
⎢                                                          c    a                                  ⎥
⎢                                                                                                  ⎥
⎢                                                         2   1                                    ⎥
⎢                                                       - ─ + ─                  2                 ⎥
⎢                                                         c   a    1    ⎛  2   1⎞    1             ⎥
⎢      0        0         0                 0         - ─────── - ────  ⎜- ─ + ─⎟  - ──      0     ⎥
⎢                                                           2        2  ⎝  c   a⎠     2            ⎥
⎢                                                          c      a⋅c                c             ⎥
⎢                                                                                                  ⎥
⎢  1   1              1   1                              1   1                                     ⎥
⎢- ─ + ─            - ─ + ─                            - ─ + ─                                    2⎥
⎢  e   a    1   1     e   a    1            1            e   a    1            1         ⎛  1   1⎞ ⎥
⎢─────── + ───  ─   ─────── + ───           ─          ─────── + ───           ─         ⎜- ─ + ─⎟ ⎥
⎣   d      a⋅d  d      d      a⋅d           d             d      a⋅d           d         ⎝  e   a⎠ ⎦

In [8]: %time ok=M2.rref()                                                                                                                     
CPU times: user 1.81 s, sys: 10.6 ms, total: 1.82 s
Wall time: 1.83 s
```
I haven't waited long enough to see that rref complete on master but it certainly is slower than 2 seconds.

The `DomainMatrix` class from #18844 can compute this rref over `QQ(a,b,c,d,e)` much faster than `Matrix` can in 1.5, 1.6 or master:
```julia
In [16]: from sympy.polys.polymatrix import DomainMatrix                                                                                       

In [17]: Mdom = DomainMatrix.from_list_sympy(M2.tolist())                                                                                      

In [18]: %time R, p = Mdom.rref()                                                                                                              
CPU times: user 38.9 ms, sys: 213 µs, total: 39.1 ms
Wall time: 39.7 ms
```
It also produces the result in a slightly simpler form in this case:
```julia
In [19]: R.to_Matrix()                                                                                                                         
Out[19]: 
⎡                           2                  2⎤
⎢    -a⋅e                - a ⋅d + 2⋅a⋅d⋅e - d⋅e ⎥
⎢1  ───────  0  0  0  0  ───────────────────────⎥
⎢   a - 2⋅e                    2          2     ⎥
⎢                             a ⋅e - 2⋅a⋅e      ⎥
⎢                                               ⎥
⎢0     0     1  0  0  0             0           ⎥
⎢                                               ⎥
⎢0     0     0  1  0  0             0           ⎥
⎢                                               ⎥
⎢0     0     0  0  1  0             0           ⎥
⎢                                               ⎥
⎢0     0     0  0  0  1             0           ⎥
⎢                                               ⎥
⎢0     0     0  0  0  0             0           ⎥
⎢                                               ⎥
⎣0     0     0  0  0  0             0           ⎦
```
This is compare to the matrix result from sympy 1.5:
```julia
In [4]: R                                                                                                                                      
Out[4]: 
⎡                                             2 ⎤
⎢                                    ⎛  1   1⎞  ⎥
⎢                                    ⎜- ─ + ─⎟  ⎥
⎢           1                        ⎝  e   a⎠  ⎥
⎢1  ─────────────────  0  0  0  0  ─────────────⎥
⎢     ⎛  1   1      ⎞                1   1      ⎥
⎢     ⎜- ─ + ─      ⎟              - ─ + ─      ⎥
⎢     ⎜  e   a    1 ⎟                e   a    1 ⎥
⎢   d⋅⎜─────── + ───⎟              ─────── + ───⎥
⎢     ⎝   d      a⋅d⎠                 d      a⋅d⎥
⎢                                               ⎥
⎢0          0          1  0  0  0        0      ⎥
⎢                                               ⎥
⎢0          0          0  1  0  0        0      ⎥
⎢                                               ⎥
⎢0          0          0  0  1  0        0      ⎥
⎢                                               ⎥
⎢0          0          0  0  0  1        0      ⎥
⎢                                               ⎥
⎢0          0          0  0  0  0        0      ⎥
⎢                                               ⎥
⎣0          0          0  0  0  0        0      ⎦
```

**Comment by oscarbenjamin** [comment]
If we can get a fix for this then I think we can make a 1.6.1 release with this and #19431

**Comment by oscarbenjamin** [comment]
Basically I think we should add a method argument to the `Matrix.exp` method giving users a few ways of calculating the matrix exponential. That would be analogous to the method argument to `Matrix.multiply` and `Matrix.pow`. I can think of three methods currently possible:

1. jordan - this would be the old method from before sympy 1.6. Basically the current method but avoids using dotprodsimp
2. jordansimp - this would be the current method with dotprodsimp
3. analytic - this would be `M.analytic_func(exp(x), x)`

**Comment by oscarbenjamin** [comment]
I can see the difficulty that @Pristine-Cat had with trying to work the dotprodsimp parameter into the matrices routines in a consistent way because there are so many inconsistent call signatures for specifying how to simplify things in the matrices code.

It's clear to me now that longer term the real fix for the problems that dotprodsimp solves is to use something like `DomainMatrix` from #18844. I'm wondering now whether we should just remove dotprodsimp altogether or disable it by default in most functions. There are certain simple cases where it helps to avoid blowup but those simpler cases are better handled by `DomainMatrix`. In the more complicated cases dotprodsimp just seems to slow things down enormously.

The example here looks like the kind of case where it might help but it really doesn't. I've tried a few heuristics to limit the slowdowns caused by dotprodsimp in the example here but I can't get it down to a reasonable time.

Removing dotprodsimp now would lead to a bunch of XFAIL tests and would effectively unclose some issues because there are things that it fixes.

**Comment by Pristine-Cat** [comment]
*cough* context manager *cough* :)

**Comment by oscarbenjamin** [comment]
We could have a context manager and a global variable. That would be a quick fix right now and is better than an environment variable.

The problem is if this is something that we end up not wanting to keep in the long term then I don't want to introduce pervasive API (environment variables, context managers, global variables) that we don't intend to honour going forwards.

Also the real question here is what is the *default*. Many people who find that matrices are slow in their application will not discover that any context manager or environment variable exists to alter that behaviour.

So if we had a context manager would it be for enabling or for disabling dotprodsimp? The conservative option to avoid regressions compared to 1.5 would be that dotprodsimp is disabled by default and that you would use the context manager to opt in.

**Comment by Pristine-Cat** [comment]
It is the least bad of several imperfect solutions. The global variable is there now it is just hidden, the context manager would just change it instead of the environment test. Tests relying on dotprodsimp would be changed to use ```with dotprodsimp(True):```. I think I could find some time this week to do this if you want to go this way.

**Comment by oscarbenjamin** [comment]
Thanks @Pristine-Cat. I think it might be the best way. Then we can consider dotprodsimp to be experimental and if a user encounters a situation that might benefit we can still suggest to use it. It can still be helpful for avoiding blowup where relevant given that `DomainMatrix` is quite a way off being available for use.

**Comment by Pristine-Cat** [comment]
See #19588

**Comment by oscarbenjamin** [comment]
Closed by #19612 and #19588

**Comment by clinssen** [comment]
@oscarbenjamin @Pristine-Cat much obliged for your exceptional response and work in general!
